### PR TITLE
DS-2750 REST API ignores expand=all on communities

### PR DESCRIPTION
**Pull request by helix84**
https://jira.duraspace.org/browse/DS-2750

### CODE REVIEW

**Comment by peterdietz** [comment]
+1 Looks good to me. (Haven't run though).

**Comment by tdonohue** [comment]
:+1:  Looks like an obvious fix to me. Probably should just be merged

File: dspace-rest/src/main/java/org/dspace/rest/CommunitiesResource.java [hunk 1]
```diff
@@ -145,7 +145,7 @@ public Community getCommunity(@PathParam("community_id") Integer communityId, @Q
 
         log.info("Reading all communities.(offset=" + offset + " ,limit=" + limit + ").");
         org.dspace.core.Context context = null;
-        ArrayList<Community> communities = null;
+        ArrayList<Community> communities = new ArrayList<Community>();
```
Comment (peterdietz) [inline dspace-rest/src/main/java/org/dspace/rest/CommunitiesResource.java]: This change doesn't do anything. the arraylist is initialized at line155. If you are trying to propogate the expand, that was deliberately omitted from being recursive. Extremely bad performance, if you attempted to pass the expand all to everything. i.e. you could get all top communities, and all of their subcommunities, and their collections and all of their items, and all of their bitstreams. Very cool, and convenient for a small instance, would crush a larger site. Perhaps you could invent a new expand option. Such as ?expand=sudo-all-yes-I-understand-the-implications

**Comment by peterdietz** [comment]
If you REALLY wanted to propogate the expand, you would have to make it so that we don't construct a sub-collection, or subcommunity with a null expand.
https://github.com/DSpace/DSpace/blob/master/dspace-rest/src/main/java/org/dspace/rest/common/Community.java#L87

https://github.com/DSpace/DSpace/blob/master/dspace-rest/src/main/java/org/dspace/rest/common/Community.java#L102

..and others

from
subcommunities.add(new Community(subCommunity, null, context));

to 
subcommunities.add(new Community(subCommunity, expand, context));

You would have to ensure you don't expand parents, otherwise that will definitely cause infinite loops.

So. I'm -1 on this PR as is. and about plus-or-minus-1 on the concept of sudo-expand

**Pull request closed without merging**